sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala

```
@@ -0,0 +1,116 @@
+        val fromJson = jsonToStructs.head.asInstanceOf[JsonToStructs].copy(schema = c.dataType)
+        val nullFields = c.children.grouped(2).flatMap {
+          case Seq(name, value) => Seq(name, Literal(null, value.dataType))
+        }.toSeq
+
+        If(IsNull(fromJson.child), c.copy(children = nullFields), KnownNotNull(fromJson))
+      } else {
+        c
+      }
+
+    case jsonToStructs @ JsonToStructs(_, options1,
+      StructsToJson(options2, child, timeZoneId2), timeZoneId1)
+        if options1.isEmpty && options2.isEmpty && timeZoneId1 == timeZoneId2 &&
+          jsonToStructs.dataType == child.dataType =>
+      // `StructsToJson` only fails when `JacksonGenerator` encounters data types it
+      // cannot convert to JSON. But `StructsToJson.checkInputDataTypes` already
+      // verifies its child's data types is convertible to JSON. But in
+      // `StructsToJson(JsonToStructs(...))` case, we cannot verify input json string
+      // so `JsonToStructs` might throw error in runtime. Thus we cannot optimize
+      // this case similarly.
+      child
+
+    case g @ GetStructField(j @ JsonToStructs(schema: StructType, _, _, _), ordinal, _)
+        if schema.length > 1 =>
+      val prunedSchema = StructType(Seq(schema(ordinal)))
+      g.copy(child = j.copy(schema = prunedSchema), ordinal = 0)
+
+    case g @ GetArrayStructFields(j @ JsonToStructs(schema: ArrayType, _, _, _), _, _, _, _)
+        if schema.elementType.asInstanceOf[StructType].length > 1 =>
+      val prunedSchema = ArrayType(StructType(Seq(g.field)), g.containsNull)
+      g.copy(child = j.copy(schema = prunedSchema), ordinal = 0, numFields = 1)
+  }
+
+  val csvOptimization: PartialFunction[Expression, Expression] = {
+    case g @ GetStructField(j @ CsvToStructs(schema: StructType, _, _, _), ordinal, _)
+        if schema.length > 1 =>
+      val prunedSchema = StructType(Seq(schema(ordinal)))
+      g.copy(child = j.copy(schema = prunedSchema), ordinal = 0)
```

@viirya, this has a behaviour about parse mode. For example, previously it worked in failfast mode when other columns can be passed but. Looks like now it fails, for example:

```scala
spark.sparkContext.parallelize(Seq("a,b")).toDF("csv")
  .selectExpr("from_csv(csv, 'a string, b string', map('mode', 'failfast')) as parsed")
  .selectExpr("parsed.a").show()
```

Before this PR:

```
+---+
|  a|
+---+
|  a|
+---+
```

After this PR:

```
org.apache.spark.SparkException: Malformed records are detected in record parsing. Parse Mode: FAILFAST. To process malformed records as null result, try setting the option 'mode' as 'PERMISSIVE'.
```

---

Good catch! `from_csv` needs to pass in both data schema and required schema for pruning usage.

---

Hm, rethinking about it, I think it is safer to do pruning only if `options` is empty. This optimization might hide a parsing error when parsing entire schema under "failfast" mode.

Although it is arguable if it still should throw the same error if we only need a field that could be parsed correctly. But I don't want to change the behavior before the optimization.

---

I re-checked Json optimization. Json parser behavior seems different than Csv here. Pruned schema doesn't change failfast behavor of json parser.